### PR TITLE
Make 'assimilation' differentiable/vectorized

### LINKED ISSUE #42: [Task]: Make `Assimilation` differentiable and efficient
The module [Assimilation ](https://github.com/ajwdewit/pcse/blob/master/pcse/crop/assimilation.py) has several classes. Only `WOFOST72_Assimilation` should be changed which is related to wofost72. See explanations and requirements in https://github.com/WUR-AI/diffWOFOST/issues/39.

### PR DESCRIPTION
**Pull request by SCiarella**
closes #42

### CODE REVIEW

File: src/diffwofost/physical_models/crop/assimilation.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+        self.params_shape = _get_params_shape(self.params)
+        self.rates = self.RateVariables(kiosk, publish=["PGASS"])
+
+        # 7-day running average buffer for TMIN (stored as tensors).
+        self._tmn_window = deque(maxlen=7)
+
+        # Reused scalar constants
+        self._epsilon = torch.tensor(1e-12, dtype=self.dtype, device=self.device)
+
+        _exist_required_external_variables(self.kiosk)
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply src/diffwofost/physical_models/crop/assimilation.py]: This check should not be here, because the `LAI` is not generated in initialization.

File: src/diffwofost/physical_models/crop/assimilation.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+        lai = _broadcast_to(k["LAI"], self.params_shape, dtype=self.dtype, device=self.device)
+
+        # Weather drivers
+        irrad = _get_drv(drv.IRRAD, self.params_shape, dtype=self.dtype, device=self.device)
+        dtemp = _get_drv(drv.DTEMP, self.params_shape, dtype=self.dtype, device=self.device)
+        tmin = _get_drv(drv.TMIN, self.params_shape, dtype=self.dtype, device=self.device)
+
+        # 7-day running average of TMIN
+        self._tmn_window.appendleft(tmin)
+        tminra = torch.stack(list(self._tmn_window), dim=0).mean(dim=0)
```
[the remark this answers is not in the document]
Comment (SCiarella) [reply src/diffwofost/physical_models/crop/assimilation.py]: thanks I totally missed that!

**Comment by SCiarella** [comment]
@SarahAlidoost thanks for the suggestions! I have implemented all of them. I think that masking the average was the crucial point, causing the failure of the tests.

File: src/diffwofost/physical_models/crop/assimilation.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    fgros = fgros * LAI
+    return fgros
+
+
+class WOFOST72_Assimilation(SimulationObject):
+    """Gross assimilation for WOFOST (potential production).
+
+    WOFOST calculates daily gross CO2 assimilation from absorbed radiation and
+    the leaf photosynthesis-light response curve. This response depends on
+    temperature and is parameterized using AFGEN tables.
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/assimilation.py]:
```suggestion
    """Class implementing a WOFOST/SUCROS style assimilation routine.

    WOFOST calculates the daily gross CO2 assimilation rate of a crop
    from the absorbed radiation and the photosynthesis-light response curve
    of individual leaves. This response is dependent on temperature and
    leaf age. The absorbed radiation is calculated from the total incoming
    radiation and the leaf area. Daily gross CO2 assimilation is obtained
    by integrating the assimilation rates over the leaf layers and over the
    day.
```

File: src/diffwofost/physical_models/crop/assimilation.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+    **Simulation parameters** (provide in cropdata dictionary)
+
+    | Name   | Description                                                        | Type | Unit                               |
+    |--------|--------------------------------------------------------------------|------|------------------------------------|
+    | AMAXTB | Max. leaf CO2 assimilation rate as function of DVS                 | TCr  | kg CO2 ha⁻¹ leaf h⁻¹               |
+    | EFFTB  | Initial light use efficiency as function of DTEMP                  | TCr  | kg CO2 ha⁻¹ h⁻¹ /(J m⁻² s⁻¹)      |
+    | KDIFTB | Extinction coefficient for diffuse visible light as function of DVS| TCr  | -                                  |
+    | TMPFTB | Reduction factor on AMAX as function of DTEMP                      | TCr  | -                                  |
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/assimilation.py]:
```suggestion
    | TMPFTB | Reduction factor on AMAX as function of daily mean temperature                      | TCr  | -                                  |
```

File: tests/physical_models/crop/test_assimilation.py [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+            self.external_states,
+            device=self.device,
+        )
+        engine.run_till_terminate()
+        results = engine.get_output()
+
+        return {"PGASS": torch.stack([item["PGASS"] for item in results])}
+
+
+def _afgen_y_mask(table_1d: torch.Tensor) -> torch.Tensor:
```
Comment (SarahAlidoost) [inline tests/physical_models/crop/test_assimilation.py]: I think this function is better to be in `utils`.
